File: react/react-native-faces-capture/ios/RNFacesCapture.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>
#import <React/RCTViewManager.h>

@interface RCT_EXTERN_MODULE(RNFacesCapture, RCTEventEmitter)

RCT_EXTERN_METHOD(startGuidedCapture:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stopGuidedCapture:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(setTargetPose:(NSString *)targetPose
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(capturePhoto:(NSString *)targetPose
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

+ (BOOL)requiresMainQueueSetup { return NO; }

@end

@interface RCT_EXTERN_MODULE(RNFacesCaptureViewManager, RCTViewManager)
@end

File: react/react-native-faces-capture/ios/RNFacesCapture.swift
import Foundation
import FacesCaptureKit

@objc(RNFacesCapture)
class RNFacesCapture: RCTEventEmitter {
    override func supportedEvents() -> [String]! {
        ["onCaptureState"]
    }

    override func startObserving() {
        FacesCaptureKit.shared.onCaptureState = { [weak self] state in
            var body: [String: Any] = [
                "targetPose": state.targetPose.rawValue,
                "faceInsideGuide": state.faceInsideGuide,
                "lightingOk": state.lightingOk,
                "sharpnessOk": state.sharpnessOk,
                "stable": state.stable,
                "canCapture": state.canCapture,
                "blockReason": state.blockReason.rawValue,
                "yaw": state.yaw,
                "verticalRatio": state.verticalRatio,
            ]
            if let detectedPose = state.detectedPose {
                body["detectedPose"] = detectedPose.rawValue
            }
            if let rect = state.faceRect {
                body["faceRect"] = [
                    "x": rect.x,
                    "y": rect.y,
                    "width": rect.width,
                    "height": rect.height,
                ]
            }
            self?.sendEvent(withName: "onCaptureState", body: body)
        }
    }

    override func stopObserving() {
        FacesCaptureKit.shared.onCaptureState = nil
    }

    @objc func startGuidedCapture(_ resolve: @escaping RCTPromiseResolveBlock,
                                  rejecter reject: @escaping RCTPromiseRejectBlock) {
        FacesCaptureKit.shared.startGuidedCapture()
        resolve(nil)
    }

    @objc func stopGuidedCapture(_ resolve: @escaping RCTPromiseResolveBlock,
                                 rejecter reject: @escaping RCTPromiseRejectBlock) {
        FacesCaptureKit.shared.stopGuidedCapture()
        resolve(nil)
    }

    @objc func setTargetPose(_ targetPose: String,
                             resolver resolve: @escaping RCTPromiseResolveBlock,
                             rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let pose = CapturePose(rawValue: targetPose) else {
            reject("INVALID_POSE", "Unknown capture pose: \(targetPose)", nil)
            return
        }
        FacesCaptureKit.shared.setTargetPose(pose)
        resolve(nil)
    }

    @objc func capturePhoto(_ targetPose: String,
                            resolver resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let pose = CapturePose(rawValue: targetPose) else {
            reject("INVALID_POSE", "Unknown capture pose: \(targetPose)", nil)
            return
        }
        do {
            let uri = try FacesCaptureKit.shared.capturePhoto(targetPose: pose)
            resolve(uri)
        } catch {
            reject("CAPTURE_NOT_IMPLEMENTED", error.localizedDescription, error)
        }
    }
}
